### ISSUE TITLE
MacOs, Safari, JS Error - TypeError: Attempted to assign to readonly property

### ISSUE THREAD

**Issue by genyslt** [calebdwilliams/element-internals-polyfill#109]
Version newest
OS: Mac
Browser: Safari

I get an JS error and nothing is working.

![image](https://user-images.githubusercontent.com/12429511/222068825-34024b9c-c54d-4b67-99d2-96a93814f4da.png)

**Comment by calebdwilliams** [comment]
I updated to Safari 16.3 (18614.4.6.1.6) this morning and the polyfill is working as expected. Are you using this polyfill in conjunction with any other polyfills (like the scoped elements polyfill for example) that might also be patching `customElements.define`?

**Comment by genyslt** [comment]
I use extra @ungap/custom-elements

But with element-internals-polyfill version 1.1.18 - there is no error.

I don't verify every is ok, but error is gone.

**Comment by calebdwilliams** [comment]
Can you see if they’re overwriting define?

**Comment by WebReflection** [comment]
> is being overly-defensive with how it treats that function and disallows other scripts from rewriting the function

funny you mention that, when the polyfill itself patches after feature detection **twice** Safari and all definitions [are configurable](https://github.com/ungap/custom-elements/blob/master/index.js#L526-L552) ... so why not using `defineProperty` there instead of shadowing an inherited method like that which *could* have potential side effects if another polyfill ahead of time define any of those methods as getters or setters? This was my thinking using `defineProperty` and I suggest this polyfill does the same ... **however** I don't mind adding `writable` in there if there's no will to use proper ECMAScript primitives to enrich native DOM utilities.

> I'm going to close this issue as wontfix as it isn't a problem with this package

It is in the sense they made a change on the way they override the native function because as you said before, it was working previous to latest updates.

I don't care discussing who should fix what as polyfills exist to enable developers, not to be on their way, but I would prefer if builtin ECMAScript stuff would, as good practice, be defined via `Object.defineProperty/ies` specially when stuff is inherited and nobody knows if a getter or setter was previously set underneath so that `customElements.define = ...` could side-effect in undesired ways.

**Comment by WebReflection** [comment]
P.S.

> however, the issue disappears if you load the element-internals-polyfill first (although it will likely introduce different errors for browsers that don't support custom elements)

the order should be logical: fix *customElements* (or bring it in) *then* fix the rest.

I am curious about what is this fix needed for `connectedCallback` as after 10 years polyfilling for production *customElements* I never ever had a bug around it ... as I don't see any feature detection there, just a replacement, can anyone please explain to me what is it that `connectedCallback` needs as fix?

Maybe this is an option to bring such fix within my polyfill and merge this one in there too as variant, or avoid re-patching anything if already patched? Just thinking out loudly here, thanks in advance for any sort of clarification.

**Comment by calebdwilliams** [comment]
Also the patch was to make sure the `formDisabledCallback` was called in the appropriate order.

**Comment by WebReflection** [comment]
> Also the patch was to make sure the formDisabledCallback was called in the appropriate order.

I see ... I don't have that as Custom Elements specs to patch so it doesn't belong to my polyfill ... fair enough, please consider using defineProperty though and everyone should be happy? Let me know if I should instead make that writable too (talking about `customElements.define` here, just to be sure we're on the same page).

**Comment by calebdwilliams** [comment]
@WebReflection I do think it still makes sense to make `customElements.define` writable for accuracy and completeness, too.

**Comment by WebReflection** [comment]
@calebdwilliams in this specific polyfill case for the DOM, where everything is enumerable and configurable down to their prototypal inheritance, I might agree using direct assignment won't be too bad, after all, actually it preserves this loop (nobody uses or needs anymore, but that's the env we work on)

```js
for (const key in customElements) console.log(key);
// define, get, whenDefined, upgrade
```

So yes, I'll make it writable or I'll drop directly all `Object.defineProperty` things around it to save code and bootstrap perf.

Maybe this conversation had the DOM caveats to consider (descriptors different from default ECMAScript world) and indeed I've made counter-patches/poly uglier for no reason (in my defence, I haven't seen one in years to be honest).

I will update the poly avoiding conflicts with this one tomorrow, you decide what to do with the rest of your code but no conflicts should happen after my next release 👋

**Comment by calebdwilliams** [comment]
@WebReflection Yeah, this polyfill is _far_ from perfect but we all just do the best we can (for free for some reason).

I’m going to update mine as well just to keep things 100. Appreciate all the OSS work you do.

**Comment by WebReflection** [comment]
FWIW I usually deliver on time but I've just realized updating that polyfill means updating other 5 projects behind and rollup breaking changes is not helping ... I'll try to land changes before next week 😅

**Comment by WebReflection** [comment]
FYI the latest custom-elements polyfill should already play better with this one 'cause every single property/field is now directly set to where it belongs.

Thanks for your patience 👋

**Comment by calebdwilliams** [comment]
This is also fixed on my end as well to prevent any future conflicts. Thanks @WebReflection and @genyslt.
